src/vs/platform/quickinput/browser/quickInput.ts

```diff
@@ -227,4 +230,75 @@ export class QuickInputService extends Themable implements IQuickInputService {
+		quickPick.onDidTriggerItemButton((e) => {
+			if (e.button.iconClass === ThemeIcon.asClassName(Codicon.pin) || e.button.iconClass !== ThemeIcon.asClassName(Codicon.pinned)) {
+				this._formatPinnedItems(storageKey, quickPick, storageService, callback, e);
+			}
+		});
+		this._formatPinnedItems(storageKey, quickPick, storageService, callback);
+	}
+
+	private async _formatPinnedItems(storageKey: string, quickPick: IQuickPick<IQuickPickItem>, storageService: IStorageService, callback?: () => Promise<void>, event?: IQuickPickItemButtonEvent<IQuickPickItem>): Promise<void> {
+		const formattedItems: (IQuickPickItem | IQuickPickSeparator)[] = [];
+		const labels = this._getPinnedLabels(storageKey, storageService);
+		const pinnedLabels = event ? this._updatePinned(storageKey, event.item.label, storageService, new Set(labels).has(event.item.label)) : labels;
+		for (const label of pinnedLabels) {
+			const item = quickPick.items.find(i => i.label === label);
+			if (item) {
+				const pinnedItem = Object.assign({}, item);
+				this._updateButtons(pinnedItem, false);
+				formattedItems.push(pinnedItem);
+			}
+		}
+		if (pinnedLabels.length) {
+			formattedItems.push({ type: 'separator', label: localize("terminal.commands.pinned", 'Pinned') });
+		}
+
+		for (const item of quickPick.items.filter(i => !!i.label)) {
+			this._updateButtons(item, true);
+			formattedItems.push(item);
+		}
+		quickPick.items = formattedItems;
+		quickPick.show();
+	}
+
+	private _updateButtons(item: IQuickPickItem | IQuickPickSeparator, removePin: boolean): void {
+		if (item.type === 'separator') {
+			return;
+		}
+		item.buttons = item.buttons ? item.buttons?.filter(b => b.iconClass !== ThemeIcon.asClassName(Codicon.pin) && b.iconClass !== ThemeIcon.asClassName(Codicon.pinned)) : [];
+		if (item.buttons.length) {
+			item.buttons.unshift({
+				iconClass: removePin ? ThemeIcon.asClassName(Codicon.pin) : ThemeIcon.asClassName(Codicon.pinned),
+				tooltip: removePin ? localize('pinCommand', "Pin command") : localize('pinnedCommand', "Pinned command"),
+				alwaysVisible: false
+			});
+		} else {
+			item.buttons.push({
+				iconClass: removePin ? ThemeIcon.asClassName(Codicon.pin) : ThemeIcon.asClassName(Codicon.pinned),
+				tooltip: removePin ? localize('pinCommand', "Pin command") : localize('pinnedCommand', "Pinned command"),
+				alwaysVisible: false
+			});
+		}
+	}
+
+	private _updatePinned(storageKey: string, label: string, storageService: IStorageService, removePin: boolean): string[] {
+		let pinnedLabels = this._getPinnedLabels(storageKey, storageService);
+		const set = new Set(pinnedLabels);
+		if (removePin) {
+			set.delete(label);
+			pinnedLabels = Array.from(set);
+		} else if (!set.has(label)) {
+			pinnedLabels.push(label);
+		}
+		storageService.store(storageKey, JSON.stringify(pinnedLabels), StorageScope.WORKSPACE, StorageTarget.USER);
```

Comment:
Instead of converting to/from set here I think a simple `Array.filter` should work?